### PR TITLE
Fix tests on windows

### PR DESCRIPTION
**Pull request by evandrocoan**
https://anki.tenderapp.com/discussions/beta-testing/1860-error-unused-type-ignore-comment

You can check individual commits for their change. Here is the list of commit messages:
1. Fixed mypy and pylint not building on Windows
    1. error: unused 'type: ignore' comment
    2. Module 'pywintypes' has no 'error' member
    3. Module 'win32pipe' has no 'SetNamedPipeHandleState' member,
       but source is unavailable. Consider adding this module to
       extension-pkg-whitelist if you want to perform analysis based on
       run-time introspection of living objects.
1. Fixed error: Incompatible types in assignment on qt/aqt/main.py
    aqt\main.py:1437: error: Incompatible types in assignment (expression has type Module, variable has type "LibraryLoader[WinDLL]")  [assignment]
                    _dummy = wintypes
                             ^
1. Fixed typecheck-setup.sh: sudo: command not found on Windows
1. Fixed make check for pylint not working on Windows due
    F:\Python\python.exe: can't open file '/cygdrive/f/anki/pyenv/Scripts/pylint': [Errno 2] No such file or directory
1. Fixed trailing-newlines.sh not failing when rg is not installed
1. Fixed test_importing.py unit tests failing on windows with
   PermissionError: [Errno 13] Permission denied
1. Fixed UnicodeEncodeError: 'charmap' codec can't encode character
    '\u2068' in position 2724: character maps to <undefined> for
    pylib/tests/test_stats.py

Here is the console output/results after running the tests now:
<details>
<p>

```
F:\anki>make check
ripgrep 11.0.2 (rev 3de31f7527)
-SIMD -AVX (compiled)
+SIMD -AVX (runtime)
make[1]: Entering directory '/cygdrive/f/anki/rslib'
cargo test --lib -- --nocapture
   Compiling anki v2.1.22 (F:\anki\rslib)
    Finished test [unoptimized + debuginfo] target(s) in 19.39s
     Running target\debug\deps\anki-1b4b4831629e4256.exe

running 35 tests
test i18n::test::numbers ... ok
test i18n::test::i18n ... ok
test cloze::test::cloze ... ok
test latex::test::latex ... ok
test media::changetracker::test::change_tracking ... ok
test media::check::test::trash_handling ... ok
test media::files::test::add_hash_suffix ... ok
test media::files::test::adding_removing ... ok
test media::database::test::database ... ok
test media::files::test::normalize ... ok
test media::files::test::truncation ... ok
test media::sync::test::required_change ... ok
test media::sync::test::sync ... ok
test sched::cutoff::test::fixed_offset ... ok
test media::check::test::unicode_normalization ... ok
test sched::cutoff::test::next_day_at ... ok
test sched::cutoff::test::rollover ... ok
test media::check::test::media_check ... ok
test sched::cutoff::test::days_elapsed ... ok
test sched::timespan::test::answer_buttons ... ok
test sched::timespan::test::combo ... ok
test template::test::alt_syntax ... ok
test sched::timespan::test::time_spans ... ok
test template::test::nonempty ... ok
test template::test::parsing ... ok
test template::test::field_empty ... ok
test template::test::render_single ... ok
test template::test::requirements ... ok
test template_filters::test::hint ... ok
test template_filters::test::tts ... ok
test template_filters::test::cloze ... ok
test template_filters::test::furigana ... ok
test template_filters::test::typing ... ok
test text::test::audio ... ok
test text::test::stripping ... ok

test result: ok. 35 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

cargo fmt -- --check
cargo clippy -- -D warnings
   Compiling anki v2.1.22 (F:\anki\rslib)
    Finished dev [unoptimized + debuginfo] target(s) in 6.95s
make[1]: Leaving directory '/cygdrive/f/anki/rslib'
make[1]: Entering directory '/cygdrive/f/anki/rspy'
touch ../proto/backend.proto
FTL_TEMPLATE_DIRS="../qt/ftl" FTL_LOCALE_DIRS="../qt/ftl/repo/desktop" \
        CARGO_TARGET_DIR="target" maturin develop
🔗 Found pyo3 bindings
🐍 Found CPython 3.8 at python.exe
   Compiling anki v2.1.22 (F:\anki\rslib)
   Compiling ankirspy v2.1.22 (F:\anki\rspy)
    Finished dev [unoptimized + debuginfo] target(s) in 22.28s
touch .build/develop
make[1]: Leaving directory '/cygdrive/f/anki/rspy'
make[1]: Entering directory '/cygdrive/f/anki/pylib'
protoc --proto_path=../proto --python_out=anki --mypy_out=anki ../proto/fluent.proto ../proto/backend.proto
Writing mypy to fluent_pb2.pyi
Writing mypy to backend_pb2.pyi
perl -i'' -pe 's/from fluent_pb2/from anki.fluent_pb2/' anki/backend_pb2.pyi
perl -i'' -pe 's/import fluent_pb2/import anki.fluent_pb2/' anki/backend_pb2.py
make[1]: Leaving directory '/cygdrive/f/anki/pylib'
make[1]: Entering directory '/cygdrive/f/anki/pylib'
python -m mypy  anki
Success: no issues found in 43 source files
python -m pytest --durations=1 -s
=========================================================================== test session starts ============================================================================
platform win32 -- Python 3.8.1, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: F:\anki\pylib
collected 129 items

tests\test_cards.py .....
tests\test_collection.py .......
tests\test_decks.py .....
tests\test_exporting.py .....
tests\test_find.py ....
tests\test_flags.py .
tests\test_importing.py ...........
tests\test_latex.py .
tests\test_media.py ...
tests\test_models.py .............
tests\test_schedv1.py ...............................
tests\test_schedv2.py .....................................
tests\test_stats.py ...
tests\test_template.py .
tests\test_undo.py ..

============================================================================= warnings summary =============================================================================
F:\anki\pyenv\lib\site-packages\win32\lib\pywintypes.py:2
  F:\anki\pyenv\lib\site-packages\win32\lib\pywintypes.py:2: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp, sys, os

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================================================================= slowest 1 test durations =========================================================================
8.14s call     tests/test_latex.py::test_latex
===================================================================== 129 passed, 1 warning in 13.34s ======================================================================
python -m black --check -t py36 anki tests setup.py tools/*.py --exclude='_pb2|buildinfo'
All done! ✨ 🍰 ✨
62 files would be left unchanged.
isort anki tests setup.py --check
Skipped 4 files
python -m pylint -j 0 --rcfile=.pylintrc -f colorized \
        --extension-pkg-whitelist=ankirspy anki tests setup.py
make[1]: Leaving directory '/cygdrive/f/anki/pylib'
make[1]: Entering directory '/cygdrive/f/anki/qt'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/cygdrive/f/anki/qt'

All checks passed!

F:\anki>
```
</p>
</details><br>

### CODE REVIEW

File: .github/scripts/trailing-newlines.sh [hunk 1]
```diff
@@ -1,8 +1,12 @@
 #!/bin/bash
 
-set -e
+set -eo pipefail
 
-files=$(rg -l '[^\n]\z' -g '!*.{svg,scss,json}' || true)
+# Checking version to force it fail the build if rg is not installed.
+# Because `set -e` does not work inside the subshell $()
+rg --version
```
Comment (dae) [inline .github/scripts/trailing-newlines.sh]: please pipe it to /dev/null

File: pylib/Makefile [hunk 2]
```diff
@@ -1,5 +1,21 @@
 SHELL := /bin/bash
 FIND := $(if $(wildcard /bin/find),/bin/find,/usr/bin/find)
+MYPY_IGNORED_IMPORTS :=
```
Comment (dae) [inline pylib/Makefile]: MYPY_ARGS might be a better, more general name

File: pylib/Makefile [hunk 3]
```diff
@@ -1,5 +1,21 @@
...
+	OS := unknown
+endif
+
+ifndef UNAME
+	UNAME := unknown
+endif
+
+# https://anki.tenderapp.com/discussions/beta-testing/1860-error-unused-type-ignore-comment
+ifneq (${OS},Windows_NT)
+	ifneq (${UNAME},Darwin)
```
Comment (dae) [inline pylib/Makefile]: I don't think you need to exclude macOS, the tests work fine here.

File: qt/Makefile [hunk 4]
```diff
@@ -1,5 +1,24 @@
...
+	OS := unknown
+endif
+
+ifndef UNAME
+	UNAME := unknown
+endif
+
+# https://anki.tenderapp.com/discussions/beta-testing/1860-error-unused-type-ignore-comment
+ifneq (${OS},Windows_NT)
+	ifneq (${UNAME},Darwin)
```
Comment (dae) [inline qt/Makefile]: same here

File: qt/Makefile [hunk 5]
```diff
@@ -1,5 +1,24 @@
 SHELL := /bin/bash
 FIND := $(if $(wildcard /bin/find),/bin/find,/usr/bin/find)
+MYPY_IGNORED_IMPORTS :=
+PYLINT_IGNORED_MODULES :=
```
Comment (dae) [inline qt/Makefile]: please use something like PYLINT_ARGS as well

File: qt/aqt/main.py [hunk 6]
```diff
@@ -1430,11 +1430,11 @@ def setupSystemSpecific(self) -> None:
             self.maybeHideAccelerators()
             self.hideStatusTips()
         elif isWin:
-            # make sure ctypes is bundled
             from ctypes import windll, wintypes  # type: ignore
 
-            _dummy = windll
-            _dummy = wintypes
+            # make sure ctypes is bundled
```
Comment (dae) [inline qt/aqt/main.py]: why did you move the comment?

File: qt/tools/typecheck-setup.sh [hunk 7]
```diff
@@ -6,8 +6,23 @@
...
+        modDir="$(cygpath -u "${modDir}")"
+        ;;
+esac
+
+cmd="rsync -a \"${TOOLS}/stubs/PyQt5/\" \"${modDir}/\""
 
-$cmd > /dev/null 2>&1 || sudo $cmd
+if [[ "w${OS}" == "wWindows_NT" ]]; 
+then
+    eval "${cmd}" > /dev/null 2>&1 || eval "${cmd}"
```
Comment (dae) [inline qt/tools/typecheck-setup.sh]: why the use of eval here?

Comment (evandrocoan) [reply qt/tools/typecheck-setup.sh]: Because it is the only way to put a command in a variable with quotes on the `rsync ` arguments. For example, before the command was like this:
```shell
cmd="rsync path1 path2"
${cmd}
```

Which works fine if and only if the `path1` and `path2` do not have spaces. If there is a space, `rsync` would receive 3 arguments due argument splitting.

To fix it, quotes need to be added on the `path1` and `path2`:
```shell
cmd="rsync \"path1\" \"path2\""
${cmd}
```

However, if you run it now the shell cannot run the command because of the expansion of the quotes. You can find some examples here: https://unix.stackexchange.com/questions/23111/what-is-the-eval-command-in-bash

File: qt/aqt/main.py (hunk 6 above)
Comment (evandrocoan) [reply qt/aqt/main.py]: I thought the `bundled` thing was happening because of the dummy variables being assignment. I moving it back.

File: qt/Makefile (hunk 5 above)
Comment (evandrocoan) [reply qt/Makefile]: Done:
```
MYPY_ARGS :=
PYLINT_ARGS :=
```

File: pylib/Makefile (hunk 3 above)
Comment (evandrocoan) [reply pylib/Makefile]: Done:
```
ifneq (${OS},Windows_NT)
	MYPY_ARGS := --warn-unused-ignores
endif
```

File: qt/Makefile (hunk 4 above)
Comment (evandrocoan) [reply qt/Makefile]: Done.

File: pylib/Makefile (hunk 2 above)
Comment (evandrocoan) [reply pylib/Makefile]: Done.

File: .github/scripts/trailing-newlines.sh (hunk 1 above)
Comment (evandrocoan) [reply .github/scripts/trailing-newlines.sh]: Done:
```shell
rg --version > /dev/null 2>&1
```

File: qt/tools/typecheck-setup.sh (hunk 7 above)
Comment (dae) [reply qt/tools/typecheck-setup.sh]: I'm familiar with eval :-) It just tends to be a red flag as it can often result in security issues:

https://stackoverflow.com/questions/17529220/why-should-eval-be-avoided-in-bash-and-what-should-i-use-instead

Perhaps a cleaner way to accomplish this would be to split the command and its arguments up into separate variables instead of trying to stuff it  all into a single variable?

Also, you're running the command 3 times instead of 2 in the non-Windows case now

Comment (dae) [reply qt/tools/typecheck-setup.sh]: (The rest of the changes look good, thanks. And a Windows CI would be nice to have as well, thank you for working on that)

Comment (evandrocoan) [reply qt/tools/typecheck-setup.sh]:
```sh
# I tried this
cmd="-a '${TOOLS}/stubs/PyQt5/' '${modDir}/'"
rsync ${cmd}

# And this:
cmd="-a \"${TOOLS}/stubs/PyQt5/\" \"${modDir}/\""
rsync ${cmd}

# But in both cases if fails with the error as originally, i.e.,
# bash still requires the eval thing to evaluate the quotes inside the variable.
+ rsync -a '"${TOOLS}/stubs/PyQt5/"' '"${modDir}/"'
+ rsync -a '"${TOOLS}/stubs/PyQt5/"' '"${modDir}/"'
rsync: change_dir "/cygdrive/f/anki/qt//"${TOOLS}/stubs/PyQt5" failed: No such file or directory (2)
rsync: change_dir#3 "/cygdrive/f/anki/qt//"${modDir}" failed: No such file or directory (2)
rsync error: errors selecting input/output files, dirs (code 3) at main.c(713) [Receiver=3.1.2]

+ rsync -a ''\''/cygdrive/f/anki/qt/tools/stubs/PyQt5/'\''' ''\''/cygdrive/f/anki/pyenv/lib/site-packages/PyQt5/'\'''
+ rsync -a ''\''/cygdrive/f/anki/qt/tools/stubs/PyQt5/'\''' ''\''/cygdrive/f/anki/pyenv/lib/site-packages/PyQt5/'\'''
rsync: change_dir "/cygdrive/f/anki/qt//'/cygdrive/f/anki/qt/tools/stubs/PyQt5" failed: No such file or directory (2)
rsync: change_dir#3 "/cygdrive/f/anki/qt//'/cygdrive/f/anki/pyenv/lib/site-packages/PyQt5" failed: No such file or directory (2)
rsync error: errors selecting input/output files, dirs (code 3) at main.c(713) [Receiver=3.1.2]
```

> Also, you're running the command 3 times instead of 2 in the non-Windows case now

I was rerunning the command because the first time its `stdout` and `stderr` were going to `/dev/null`, then, on error I rerun a second time to log out what error was happening before going into sudo mode.

I just pushed this, removing the variable and duplicating the command:
```sh
if [[ "w${OS}" == "wWindows_NT" ]];
then
    rsync -a "${TOOLS}/stubs/PyQt5/" "${modDir}/"
else
    rsync -a "${TOOLS}/stubs/PyQt5/" "${modDir}/" || sudo rsync -a "${TOOLS}/stubs/PyQt5/" "${modDir}/"
fi
```

I removed the `/dev/null` thing because here `rsync` was already not outputting anything to `stderr/stdout`. `rsync` would just be quite verbose if we were using the `-v` flag. I just tested here both on Windows and Linux and rsync is not outputting anything when there are no errors:
image

So, removing the `/dev/null` speeds things up when there is some error on `rsync` command, i.e., I would not have to push or edit the file removing it to figure out why `rsync` would not be working in case of some error.

**Comment by evandrocoan** [comment]
This needs to be merged before I can open the pull request for the GitHub Actions for Windows.

**Comment by dae** [comment]
Thanks!
